### PR TITLE
TRPL: unsafe -> raw pointers

### PR DESCRIPTION
**Pull request by steveklabnik**
Many unsafe features are now in the unstable section, so this section is
really just about raw pointers now. That also makes sense for its place
in the TOC.

This addresses part of #12905.

### CODE REVIEW

File: src/doc/trpl/raw-pointers.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+```rust
+let x = 5;
+let raw = &x as *const i32;
+
+let points_at = unsafe { *raw };
+
+println!("raw points at {}", points_at);
+```
+
+For more operations on raw pointers, see [ther API documentation][rawapi].
```
Comment (petrochenkov) [inline src/doc/trpl/raw-pointers.md]: ther -> their

File: src/doc/trpl/raw-pointers.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+there are two types that are extra-special. Much of Rust’s safety comes from
+compile-time checks, but raw pointers don’t have such guarantees, and are
+[unsafe][unsafe] to use.
+
+`*const T` and `*mut T` are called ‘raw pointers’ in Rust. Sometimes, when
+writing certain kinds of libraries, you’ll need to get around Rust’s safety
+guarantees for some reason. In this case, you can use raw pointers to implement
+your library, while exposing a safe interface for your users. For example, `*`
+pointers are allowed to alias, allowing them to be used to write
+shared-ownership types like reference counted and garbage collected pointers,
```
Comment (apasel422) [inline src/doc/trpl/raw-pointers.md]: s/reference counted/reference-counted/
s/garbage collected/garbage-collected/

File: src/doc/trpl/raw-pointers.md (hunk 1 above)
Comment (steveklabnik) [inline src/doc/trpl/raw-pointers.md]: nice catch

File: src/doc/trpl/raw-pointers.md (hunk 2 above)
Comment (steveklabnik) [inline src/doc/trpl/raw-pointers.md]: nice, i've just cleaned this up

File: src/doc/trpl/unsafe-code.md [hunk 3]
```diff
@@ -1,82 +1,4 @@
...
-let ref_2: &mut u8 = unsafe { mem::transmute(&mut *ref_1) };
-
-// oops, ref_1 and ref_2 point to the same piece of data (x) and are
-// both usable
-*ref_1 = 10;
-*ref_2 = 20;
-```
-
-## Raw pointers
+% Raw Pointers
```
Comment (alexcrichton) [inline src/doc/trpl/unsafe-code.md]: Was this file intended for deletion?

**Comment by alexcrichton** [comment]
r=me if unsafe-code.md should be deleted, and also r=me if it shouldn't :)

File: src/doc/trpl/unsafe-code.md (hunk 3 above)
Comment (steveklabnik) [inline src/doc/trpl/unsafe-code.md]: Yes, it effectively got moved to the other one. I'm working on a new, general 'unsafe' guide now, was just gonna send a different PR.

File: src/doc/trpl/raw-pointers.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
+% Raw Pointers
+
+Rust has a number of different smart pointer types in its standard library, but
+there are two types that are extra-special. Much of Rust’s safety comes from
```
Comment (huonw) [inline src/doc/trpl/raw-pointers.md]: Which two? `&` and `&mut`? or `*const` and `*mut`? It's seems to me that the raw pointer types are the least special (at least when compared to C/C++/other languages).

Also, I wouldn't call the raw pointer types "smart pointers".
